### ISSUE TITLE
[SYCL][CUDA] fgpu-rdc, Kernel Error: CUDA_ERROR_INVALID_IMAGE, using parallel_for_work_group

### ISSUE THREAD

**Issue by abagusetty** [intel/llvm#8690]
**Describe the bug**
Using `-fgpu-rdc` is triggering this error.  Thanks to the reproducer from @nscottnichols

To compile: 
```bash
clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda -Xcuda-ptxas --maxrregcount=255 -Xcuda-ptxas --verbose -Xsycl-target-backend --cuda-gpu-arch=sm_80 -fgpu-rdc sycl_fgpurdc.cpp -o sycl_fgpurdc.out
```

**Error**
```bash
PI CUDA ERROR:
	Value:           200
	Name:            CUDA_ERROR_INVALID_IMAGE
	Description:     device kernel image is invalid
	Function:        build_program
	Source Location: /..../llvm/sycl/plugins/cuda/pi_cuda.cpp:776


PI CUDA ERROR:
	Value:           400
	Name:            CUDA_ERROR_INVALID_HANDLE
	Description:     invalid resource handle
	Function:        cuda_piProgramRelease
	Source Location: /..../llvm/sycl/plugins/cuda/pi_cuda.cpp:3604

terminate called after throwing an instance of 'sycl::_V1::compile_program_error'
  what():  The program was built for 1 devices
Build program log for 'NVIDIA A100-PCIE-40GB':
 -999 (Unknown PI error)
Aborted
```
**To Reproduce**
```c++
#include <sycl/sycl.hpp>

int main(int, char**) {
  float h_a[4] { 1.0, 2.0, 3.0, 4.0 };
  float h_b[4] { 5.0, 6.0, 7.0, 8.0 };
  float h_c[4];

  sycl::queue my_queue(sycl::gpu_selector_v);

  // create and allocate device objects
  float *d_a = sycl::malloc_device<float>(4, my_queue);
  float *d_b = sycl::malloc_device<float>(4, my_queue);
  float *d_c = sycl::malloc_device<float>(4, my_queue);

  // copy from host to device
  my_queue.memcpy(d_a, h_a, 4*sizeof(float));
  my_queue.memcpy(d_b, h_b, 4*sizeof(float));
  // wait for asynchronous copies to complete
  my_queue.wait();

  my_queue.submit([&](sycl::handler& cgh) {
      cgh.parallel_for_work_group(sycl::range<1>{2}, sycl::range<1>{2}, ([=](sycl::group<1> wGroup) {
          wGroup.parallel_for_work_item([&](sycl::h_item<1> index) {
              size_t idx = index.get_global_id(0);
              // perform vector addition according to own
              // rank number starting at 0
              d_c[idx] = d_a[idx] + d_b[idx];
          });
      }));
  });
  my_queue.wait(); // wait for kernel to complete

  // copy results from device to host
  my_queue.memcpy(h_c, d_c, 4*sizeof(float)).wait();

  // free memory
  sycl::free(d_a, my_queue);
  sycl::free(d_b, my_queue);
  sycl::free(d_c, my_queue);

  for (size_t i=0; i < 4; i++) {
    std::cout << h_c[i] << std::endl;
  }
  return 0;
}

```
**Environment (please complete the following information):**

- OS: Linux
- Target device and vendor: Nvidia A100
- DPC++ version: 
```bash
$ clang -v
clang version 17.0.0 (git@github.com:intel/llvm.git 9596ea0ddd362165cefd419e57cf2b57f5962c6c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /gpfs/jlse-fs0/users/abagusetty/compilers/llvm/build_cuda_dbg/bin
Found candidate GCC installation: /soft/compilers/gcc/11.1.0/x86_64-suse-linux/lib/gcc/x86_64-pc-linux-gnu/11.1.0
Selected GCC installation: /soft/compilers/gcc/11.1.0/x86_64-suse-linux/lib/gcc/x86_64-pc-linux-gnu/11.1.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /soft/compilers/cuda/cuda-11.6.2, version 11.6
```

**Additional context**
Works fine with L0 (PVC) and HIP (MI250x)

**Comment by abagusetty** [comment]
@sarnex Any suggestions related to this

**Comment by sarnex** [comment]
I was able to reproduce the issue on A100. Let me see what's going on.

**Comment by sarnex** [comment]
I got a different bad commit:

```
commit 230d5627b017fec94b7715a3554e270ba6dfe4b0 (HEAD, refs/bisect/bad)
Merge: 6077e8eded62 8bfee2460371
Author: bb-sycl <bb-sycl@intel.com>
Date:   Tue Mar 7 20:58:21 2023 -0800

    LLVM and SPIRV-LLVM-Translator pulldown (WW08)

    LLVM: llvm/llvm-project@890146b
    SPIRV-LLVM-Translator: KhronosGroup/SPIRV-LLVM-Translator@4af277c
```

My bisect didn't look into merge commits, tomorrow I'll try to find which it was.

**Comment by abagusetty** [comment]
Thanks so much for tracking. My bisect pointed me to the same day Mar 7 range of commit as culprits. I manually narrow it down to this: 
good: e87b45680222ee3bcda7a00dc0b016f050075f10
bad: f66770956fe4000a07979a3b99773772b2980507
as if I understand things there was a bit of merge from upstream llvm already happened with your bad commit and that changed the version of clang-16 to clang-17. My bad commit me to the possible culprit even before the sync with upstream.

**Comment by sarnex** [comment]
That's weird. I just manually confirmed the below:

good: 6077e8eded6250937dd437efbb334fa83006ce10
bad: 230d5627b017fec94b7715a3554e270ba6dfe4b0 

Note bad is 1 commit after good.

f66770956fe4000a07979a3b99773772b2980507 is good for me.


Do you mind double checking the results on your side?

**Comment by abagusetty** [comment]
> 

Sure, will keep you posted.

**Comment by abagusetty** [comment]
@sarnex I can confirm that the first bad commit is indeed same as yours: `https://github.com/intel/llvm/commit/230d5627b017fec94b7715a3554e270ba6dfe4b0` 
Sorry about the false info, some how did a clean build and the above first bad commit is same as yours.

**Comment by sarnex** [comment]
@abagusetty Cool thanks for confirming, yeah I think cmake incremental builds don't work well with git bisect so I always do full builds every time. 

I'll try to find the bad commit today.

**Comment by bader** [comment]
Tagging @npmiller and @intel/llvm-reviewers-cuda for awareness.

**Comment by GeorgeWeb** [comment]
Investigating this further lead to a few outcomes and questions.

`-fgpu-rdc` requires the cuda toolchain to call ptxas with `-c` in order to generate relocatable device code binaries. This means, as expected by the behaviour of `-c` (`--compile-only`), that we need to call a device linker to link the binary. Previously, without passing `-c` everything was fine because ptxas was outputing a statically linked elf.

ptxas `-c`:
```bash
output-file.cubin: ELF 64-bit LSB relocatable, NVIDIA CUDA architecture,, not stripped
```

ptxas no `-c`:
```bash
output-file.cubin: ELF 64-bit LSB executable, NVIDIA CUDA architecture,, statically linked, not stripped
```

Now, there are things that can be done such as introducing an extra link step in the driver or maybe preferrable linking in the cuda plugin (ur adapter) via the driver program linking APIs. However, I am unsure what is the actual expectation of `-fgpu-rdc` here? Does it not make more sense to call clang++ again on the unlinked binaries, which very similar to what the flow is with `nvcc`. Also a related question - what is `-fsycl-rdc`? How does it tie to the latter as an alias when the documentation seem to explain it a bit different from cuda's rdc (or `-fgpu-rdc`).

As a proposal would it make sense to leave `-fgpu-rdc` as is and make `-fsycl-rdc` work in one `clang++` invocation? As in adding the stage to link the cubin produced from ptxas either in the runtime or the driver. And if we want to do that, do we even need `-c` passed in ptxas anymore? My point is what is the use-case and how is relocatable device code used so we can call a decision on how to proceed with addressing this issue. 

Many thanks for any possible feedback. @sarnex @mdtoguchi 

- Georgi

**Comment by sarnex** [comment]
@GeorgeWeb `-fsycl-rdc` is an alias to `-fgpu-rc` so they (should) represent the same idea. The option being enabled allows cross-TU calls, they option being disabled forbids cross-TU calls. I would say the user shouldn't have to call clang again if it is being called currently with linking enabled. If they compile with `clang++ -c` or `clang++ -fsycl-device-only`, that's a different story but I don't think that's happening here. It sounds like we need to modify the driver to call the PTX device linker if `-fgpu-rdc`/`-fsycl-rdc` is enabled. I don't think we should do it at runtime if we don't need to.

How does SYCL targeting PTX ever work right now? Does it happen to work if you only compile code with no cross-TU calls?

If you want to make cross-TU calls with CUDA, it seems you need to pass `-c` to `ptsax` and also call the device linker (which we don't do today based on your description).

**Comment by abagusetty** [comment]
Just to confirm, things are fine with the HIP adaptor, tested on MI250.

**Comment by GeorgeWeb** [comment]
Closing as https://github.com/intel/llvm/pull/11420 has already fixed this.
